### PR TITLE
Docs: Add a book preprocessor

### PR DESCRIPTION
**Pull request by Lorak-mmk**
This PR adds a preprocessor for the book that removes Sphinx-related sections.
This was previously done by `docs/build_book.py` script, which this PR removes.
There are some advantages of new approach:
- No unnecessary file writing. Book content is passed to preprocessor by stdin. This is both faster and more friendly to disk.
- No way to accidentally build a book without preprocessing. Also now developer doesn't need to know about the script or Makefile target that uses it.
- New script is simpler as it doesn't need to handle files anymore. As a contributor I can certify that it's possible to not know about the script or Makefile target and be frustrated about Sphinx sections.

There is one disadvantage: book must be built either from main directory, or docs directory. If it's built from different location, build will succeed buy Sphinx sections won't be stripped.
This is because of mdbook limitation - PWD of build process is the folder the command is executed in, not the folder in which book.toml is located.
This shouldn't be a problem - I suspect book build is never / very rarely executed from other dir.
If it becomes a problem, it's possible to fix it. The script in bash.sh will need to be modified:
- if executed with `supports` argument - just exit.
- Otherwise, read stdin using jq. From first element extract `root` element, `cd` to it's value.
- Execute `python ./sphinx_preprocessor.py` with previosuly-read stdin.

## Pre-review checklist

- [ ] I have split my patch into logically separate commits.
- [ ] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [ ] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
- [ ] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [ ] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by Lorak-mmk** [comment]
Submitted an issue to mdbook: https://github.com/rust-lang/mdBook/issues/2289

**Comment by Lorak-mmk** [comment]
mdbook issue was closed in favor of https://github.com/rust-lang/mdBook/issues/1424 which is open and not touches for 3 years now, so we shouldn't hope for mdbook fix. Imo the script I wrote should be enough, if it isn't I can write more complicated one that would always work.

File: docs/sphinx_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import json
```
Comment (piodul) [inline docs/sphinx_preprocessor.py:1]: Nit: how about adding a shebang? We had it in the previous script.

Comment (Lorak-mmk) [reply docs/sphinx_preprocessor.py:1]: This script is not supposed to be run manually, so I don't see how shebang helps?

**Comment by Lorak-mmk** [comment]
v2
- Changed contributing.md: added a note about limitations
- Removed unnecessary in the script coment.

File: docs/sphinx_preprocessor.py (hunk 1 above)
Comment (piodul) [reply docs/sphinx_preprocessor.py:1]: Makes sense not to do it.

**Review by piodul — APPROVED** [review]

**Pull request merged**